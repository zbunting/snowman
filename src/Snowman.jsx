import { useState } from "react";
import "./Snowman.css";
import img0 from "./0.png";
import img1 from "./1.png";
import img2 from "./2.png";
import img3 from "./3.png";
import img4 from "./4.png";
import img5 from "./5.png";
import img6 from "./6.png";
import { randomWord, ENGLISH_WORDS } from "./words";
import Button from "./Button.jsx";


/** Snowman game: plays hangman-style game with a melting snowman.
 *
 * Props:
 * - maxWrong: how many wrong moves is a player allowed?
 * - images: array of images for wrong guess
 * - words: array of words to pick answer from
 *
 * State:
 * - nWrong: # wrong guesses so far
 * - guessedLetters: set of guessed letters (good and bad) so far
 * - answer: selected secret word*
 */

function Snowman({
  images = [img0, img1, img2, img3, img4, img5, img6],
  words = ENGLISH_WORDS,
  maxWrong = 6,
}) {
  /** by default, allow 6 guesses and use provided gallows images. */

  const [nWrong, setNWrong] = useState(0);
  const [guessedLetters, setGuessedLetters] = useState(() => new Set());
  const [answer, setAnswer] = useState(randomWord(words));

  /** guessedWord: show current-state of word:
   if guessed letters are {a,p,e}, show "app_e" for "apple"
   */
  function guessedWord() {
    return answer
      .split("")
      .map(ltr => (guessedLetters.has(ltr) ? ltr : "_"));
  }

  /** handleGuess: handle a guessed letter:
   - add to guessed letters
   - if not in answer, increase number-wrong guesses
   */
  function handleGuess(evt) {
    let ltr = evt.target.value;

    setGuessedLetters(g => {
      const newGuessed = new Set(g);
      newGuessed.add(ltr);
      return newGuessed;
    });

    setNWrong(n => n + (answer.includes(ltr) ? 0 : 1));
  }

  /** generateButtons: return array of letter buttons to render */
  function generateButtons() {
    return "abcdefghijklmnopqrstuvwxyz".split("").map(ltr => (
      <button
        key={ltr}
        className={ltr}
        value={ltr}
        onClick={handleGuess}
        disabled={guessedLetters.has(ltr)}
      >
        {ltr}
      </button>
    ));
  }

  /** restartGame: reset game states for a new game. */
  function restartGame() {
    setNWrong(0);
    setGuessedLetters(new Set());
    setAnswer(randomWord(words));
  }

  const won = guessedWord().join("") === answer;

  return (
    <div className="Snowman">
      <img src={(images)[nWrong]} alt={nWrong} />
      <p>Number wrong: {nWrong}</p>
      <p className="Snowman-word">{guessedWord()}</p>
      {nWrong < maxWrong && !won && <p>{generateButtons()}</p>}
      {nWrong >= maxWrong && <p>You lose! The word was {answer}.</p>}
      {won && <p>You win!</p>}
      <Button label="Start a new ☃️ game!" click={restartGame} />
    </div>
  );
}


export default Snowman;
