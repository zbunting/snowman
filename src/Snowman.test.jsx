import { test, describe, expect } from "vitest";
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Snowman from "./Snowman.jsx";

const TEST_WORD = "apple";

describe("test Snowman compenent", function () {
  test("matches snapshot", function () {
    const { container } = render(<Snowman words={[TEST_WORD]} />);
    expect(container).toMatchSnapshot();
  });

  test("no button area and shows win message on win", function () {
    const { container, debug } = render(<Snowman
      words={[TEST_WORD]}
      maxWrong={2}
    />);

    fireEvent.click(container.querySelector("button[value='a']"));

    fireEvent.click(container.querySelector("button[value='p']"));
    fireEvent.click(container.querySelector("button[value='l']"));
    fireEvent.click(container.querySelector("button[value='e']"));

    expect(
      container.querySelector("button[value='a']")
    ).not.toBeInTheDocument();

    expect(
      container).toContainHTML(`You win!`);
  });

  test("image stays the same after correct guess", function () {
    const { container, debug } = render(<Snowman
      words={[TEST_WORD]}
      maxWrong={2}
    />);

    fireEvent.click(container.querySelector("button[value='a']"));

    expect(
      container.querySelector("img[alt='0']")
    ).toBeInTheDocument();
  });

  test("no button area and shows loss message on loss", function () {
    const { container, debug } = render(<Snowman
      words={[TEST_WORD]}
      maxWrong={2}
    />);

    fireEvent.click(container.querySelector("button[value='x']"));

    fireEvent.click(container.querySelector("button[value='q']"));

    expect(
      container.querySelector("button[value='a']")
    ).not.toBeInTheDocument();

    expect(
      container).toContainHTML(`You lose! The word was ${TEST_WORD}.`);
  });

  test("image changes after incorrect guess", function () {
    const { container, debug } = render(<Snowman
      words={[TEST_WORD]}
      maxWrong={2}
    />);

    expect(
      container.querySelector("img[alt='0']")
    ).toBeInTheDocument();

    fireEvent.click(container.querySelector("button[value='x']"));

    expect(
      container.querySelector("img[alt='1']")
    ).toBeInTheDocument();
  });
});