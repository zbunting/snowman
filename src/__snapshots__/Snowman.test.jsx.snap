// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <div
    class="Snowman"
  >
    <img
      alt="0"
      src="/src/0.png"
    />
    <p>
      Number wrong: 
      0
    </p>
    <p
      class="Snowman-word"
    >
      _
      _
      _
      _
      _
    </p>
    <p>
      <button
        value="a"
      >
        a
      </button>
      <button
        value="b"
      >
        b
      </button>
      <button
        value="c"
      >
        c
      </button>
      <button
        value="d"
      >
        d
      </button>
      <button
        value="e"
      >
        e
      </button>
      <button
        value="f"
      >
        f
      </button>
      <button
        value="g"
      >
        g
      </button>
      <button
        value="h"
      >
        h
      </button>
      <button
        value="i"
      >
        i
      </button>
      <button
        value="j"
      >
        j
      </button>
      <button
        value="k"
      >
        k
      </button>
      <button
        value="l"
      >
        l
      </button>
      <button
        value="m"
      >
        m
      </button>
      <button
        value="n"
      >
        n
      </button>
      <button
        value="o"
      >
        o
      </button>
      <button
        value="p"
      >
        p
      </button>
      <button
        value="q"
      >
        q
      </button>
      <button
        value="r"
      >
        r
      </button>
      <button
        value="s"
      >
        s
      </button>
      <button
        value="t"
      >
        t
      </button>
      <button
        value="u"
      >
        u
      </button>
      <button
        value="v"
      >
        v
      </button>
      <button
        value="w"
      >
        w
      </button>
      <button
        value="x"
      >
        x
      </button>
      <button
        value="y"
      >
        y
      </button>
      <button
        value="z"
      >
        z
      </button>
    </p>
    <button
      class="Button"
    >
      Start a new ☃️ game!
    </button>
  </div>
</div>
`;

exports[`test Snowman compenent > matches snapshot 1`] = `
<div>
  <div
    class="Snowman"
  >
    <img
      alt="0"
      src="/src/0.png"
    />
    <p>
      Number wrong: 
      0
    </p>
    <p
      class="Snowman-word"
    >
      _
      _
      _
      _
      _
    </p>
    <p>
      <button
        value="a"
      >
        a
      </button>
      <button
        value="b"
      >
        b
      </button>
      <button
        value="c"
      >
        c
      </button>
      <button
        value="d"
      >
        d
      </button>
      <button
        value="e"
      >
        e
      </button>
      <button
        value="f"
      >
        f
      </button>
      <button
        value="g"
      >
        g
      </button>
      <button
        value="h"
      >
        h
      </button>
      <button
        value="i"
      >
        i
      </button>
      <button
        value="j"
      >
        j
      </button>
      <button
        value="k"
      >
        k
      </button>
      <button
        value="l"
      >
        l
      </button>
      <button
        value="m"
      >
        m
      </button>
      <button
        value="n"
      >
        n
      </button>
      <button
        value="o"
      >
        o
      </button>
      <button
        value="p"
      >
        p
      </button>
      <button
        value="q"
      >
        q
      </button>
      <button
        value="r"
      >
        r
      </button>
      <button
        value="s"
      >
        s
      </button>
      <button
        value="t"
      >
        t
      </button>
      <button
        value="u"
      >
        u
      </button>
      <button
        value="v"
      >
        v
      </button>
      <button
        value="w"
      >
        w
      </button>
      <button
        value="x"
      >
        x
      </button>
      <button
        value="y"
      >
        y
      </button>
      <button
        value="z"
      >
        z
      </button>
    </p>
    <button
      class="Button"
    >
      Start a new ☃️ game!
    </button>
  </div>
</div>
`;
